modules/regexp-parser/regexp-parser.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2021 Balabit
+ * Copyright (c) 2021 Xiaoyu Qiu
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "regexp-parser.h"
+#include "parser/parser-expr.h"
+#include "scratch-buffers.h"
+#include "string-list.h"
+
+#include <string.h>
+#include <stdio.h>
+
+typedef struct _RegexpParser
+{
+  LogParser super;
+  gchar *prefix;
+  GList *patterns;
+  LogMatcherOptions matcher_options;
+  GList *matchers;
+} RegexpParser;
+
+LogMatcherOptions *
+regexp_parser_get_matcher_options(LogParser *s)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  return &self->matcher_options;
+}
+
+void
+regexp_parser_set_prefix(LogParser *s, const gchar *prefix)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  g_free(self->prefix);
+  if (prefix)
+    self->prefix = g_strdup(prefix);
+  else
+    self->prefix = NULL;
+}
+
+void
+regexp_parser_set_patterns(LogParser *s, GList *patterns)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  string_list_free(self->patterns);
+  self->patterns = patterns;
+}
+
+gboolean
+regexp_parser_compile(LogParser *s)
+{
+  RegexpParser *self = (RegexpParser *) s;
+
+  log_matcher_options_init(&self->matcher_options);
+  gboolean result = TRUE;
+
+  for (GList *item = self->patterns; item; item = item->next)
+    {
+      self->matchers = g_list_prepend(self->matchers, log_matcher_new(&self->matcher_options));
+      log_matcher_pcre_set_prefix((LogMatcher *)self->matchers->data, self->prefix);
+
+      if(!log_matcher_compile((LogMatcher *)self->matchers->data, item->data, NULL))
```

Comment:
Please handle the error in a more user-friendly way, I got no reason why the compilation of my regexp failed:

Error parsing regexp-parser, compiling the regexp failed in etc/syslog-ng-re.conf:5:11-5:61:
1       @version: 3.32
2       
3       log { 
4       	source { tcp(port(2000)); };
5-----> 	parser { regexp-parser(patterns("(?foo)a+") prefix(".re.")); };
5-----> 	         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
6       	destination { file("/tmp/log" template("$(format-json *)\n")); };
7       };

Reply:
I have updated the `regexp-parser-grammar.ym` file and the `regexp_parser_compile` function. In the latest commits, syslog-ng will output the error with the failure reason:
```
Error parsing regexp-parser, compiling the regexp failed, error=Failed to compile PCRE expression >>>(?foo)a+<<< `unrecognized character after (? or (?-' at character 2 in /etc/syslog-ng.conf:57:11-57:61:
55      log { 
56      	source(s_file);
57----> 	parser { regexp-parser(patterns("(?foo)a+") prefix(".re.")); };
57----> 	         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
58      	destination { file("/tmp/log" template("$(format-json *)\n")); };
59      };
```